### PR TITLE
feat(raredisease) add metrics-deliver

### PR DESCRIPTION
**Pull request by peterpru**
## Description
This PR aims to add `cg workflow raredisease metrics-deliver case_id`. It parses the multiqc-data.json file and creates a metrics_deliverables.yaml file. At the same time it evaluates whether the QC checks pass.

### Added

- cg workflow raredisease metrics-deliver case_id

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b add-raredisease-metrics-deliver
    ```

### How to test

- [x] Do `cg workflow raredisease metrics-deliver --help`
- [x] Do `cg workflow raredisease metrics-deliver funkitten` with the median target coverage being 25
- [x] Do `cg workflow raredisease metrics-deliver funkitten` with the median target coverage being 27

### Expected test outcome

- [x] Check that the command is working when doing --help
- [x] Check that it fails when median target coverage metric does not meet QC values
image

- [x] Check that it succeeds when the metrics do meet QC thresholds:
image

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/meta/workflow/raredisease.py [hunk 1]
```diff
@@ -190,3 +193,33 @@ def get_gene_panel(self, case_id: str) -> list[str]:
...
+        """Return a list of the metrics specified in a MultiQC json file for the case samples."""
+        multiqc_json: list[dict] = MultiqcDataJson(
+            **read_json(file_path=self.get_multiqc_json_path(case_id=case_id))
+        ).report_general_stats_data
+        samples: list[Sample] = self.status_db.get_samples_by_case_id(case_id=case_id)
+        metrics_list: list[MetricsBase] = []
+        for sample in samples:
+            sample_id: str = sample.internal_id
+            metrics_values: dict = self.parse_multiqc_json_for_sample(
+                sample_name=sample.name, multiqc_json=multiqc_json
```
Comment (fevac) [inline cg/meta/workflow/raredisease.py:207]:
```suggestion
                sample_name=sample_id, multiqc_json=multiqc_json
```

File: cg/meta/workflow/raredisease.py [hunk 2]
```diff
@@ -190,3 +193,33 @@ def get_gene_panel(self, case_id: str) -> list[str]:
...
+            metrics_list.extend(metric_base_list)
+        return metrics_list
+
+    @staticmethod
+    def parse_multiqc_json_for_sample(sample_name: str, multiqc_json: list[dict]) -> dict:
+        """Parse a multiqc_data.json and returns a dictionary with metric name and metric values for each sample."""
+        metrics_values: dict = {}
+        for stat_dict in multiqc_json:
+            for sample_key, sample_values in stat_dict.items():
+                if sample_key == f"{sample_name}_{sample_name}":
```
Comment (fevac) [inline cg/meta/workflow/raredisease.py:221]:
```suggestion
               if sample_name in sample_key:
```

File: tests/cli/workflow/raredisease/test_cli_raredisease_metrics_deliver.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+def test_metrics_deliver(
+    cli_runner: CliRunner,
+    raredisease_context: CGConfig,
+    raredisease_mock_analysis_finish,
+    caplog: LogCaptureFixture,
+    raredisease_case_id: str,
+    raredisease_metrics_deliverables_path: Path,
+):
+    """Test command with a case id and a finished analysis which should execute successfully."""
+    caplog.set_level(logging.INFO)
```
Comment (fevac) [inline tests/cli/workflow/raredisease/test_cli_raredisease_metrics_deliver.py:24]:
```suggestion
    caplog.set_level(logging.DEBUG)
```

File: tests/cli/workflow/raredisease/test_cli_raredisease_metrics_deliver.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+from cg.cli.workflow.raredisease.base import metrics_deliver
+from cg.constants import EXIT_SUCCESS
+from cg.models.cg_config import CGConfig
+from cg.io.yaml import read_yaml
+
+
+def test_metrics_deliver(
+    cli_runner: CliRunner,
+    raredisease_context: CGConfig,
+    raredisease_mock_analysis_finish,
```
Comment (fevac) [inline tests/cli/workflow/raredisease/test_cli_raredisease_metrics_deliver.py:18]: implement this one

File: tests/fixtures/analysis/raredisease/multiqc_data.json [hunk 5]
```diff
@@ -0,0 +1,24 @@
+{
+    "report_data_sources": {},
+    "report_general_stats_data": [
+        {
+            "a_sample_name_a_sample_name_raw_1": {
```
Comment (fevac) [inline tests/fixtures/analysis/raredisease/multiqc_data.json:5]:
```suggestion
            "ADM1_1": {
```

File: tests/fixtures/analysis/raredisease/multiqc_data.json [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+                "percent_duplicates": 9.864942996595772
+            }
+        },
+        {
+            "a_sample_name_a_sample_name": {
```
Comment (fevac) [inline tests/fixtures/analysis/raredisease/multiqc_data.json:10]:
```suggestion
            "ADM1": {
```

File: tests/fixtures/analysis/raredisease/multiqc_data.json [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+                "PCT_PF_READS_ALIGNED": 0.99966,
+                "MEDIAN_TARGET_COVERAGE": 28.0,
+                "PCT_TARGET_BASES_10X": 0.975685
+            }
+        }
+    ],
+    "report_saved_raw_data": {
+        "multiqc_picard_AlignmentSummaryMetrics": {
+            "a_sample_name_a_sample_name": {
```
Comment (fevac) [inline tests/fixtures/analysis/raredisease/multiqc_data.json:19]:
```suggestion
            "ADM1": {
```

File: tests/conftest.py [hunk 8]
```diff
@@ -2406,6 +2406,38 @@ def mock_config(raredisease_dir: Path, raredisease_case_id: str) -> None:
...
+@pytest.fixture(scope="function")
+def raredisease_metrics_deliverables_path(raredisease_dir: Path, raredisease_case_id: str) -> Path:
+    """Path to deliverables file."""
+    return Path(
+        raredisease_dir, raredisease_case_id, f"{raredisease_case_id}_metrics_deliverables"
+    ).with_suffix(FileExtensions.YAML)
+
+
+@pytest.fixture(scope="function")
+def raredisease_mock_analysis_finish(
```
Comment (fevac) [inline tests/conftest.py:2426]: implement this one

File: cg/meta/workflow/raredisease.py [hunk 9]
```diff
@@ -190,3 +193,33 @@ def get_gene_panel(self, case_id: str) -> list[str]:
     def get_managed_variants(self) -> list[str]:
         """Create and return the managed variants."""
         return self._get_managed_variants(genome_build=GENOME_BUILD_37)
+
+    def get_multiqc_json_metrics(self, case_id: str) -> list[MetricsBase]:
+        """Return a list of the metrics specified in a MultiQC json file for the case samples."""
+        multiqc_json: list[dict] = MultiqcDataJson(
+            **read_json(file_path=self.get_multiqc_json_path(case_id=case_id))
+        ).report_general_stats_data
+        samples: list[Sample] = self.status_db.get_samples_by_case_id(case_id=case_id)
```
Comment (fevac) [inline cg/meta/workflow/raredisease.py:202]: Sample is missing import

**Review by fevac — COMMENTED** [review]
only few things to make it functional

File: cg/meta/workflow/raredisease.py (hunk 9 above)
Comment (peterpru) [reply cg/meta/workflow/raredisease.py:202]: thanks, added

File: tests/conftest.py (hunk 8 above)
Comment (peterpru) [reply tests/conftest.py:2426]: thanks, tests passing now 😄

**Comment by peterpru** [comment]
I am missing something. I get back a file that does not have the correct contents. @fevac and @rannick , do you know what is missing? All pytests pass.

image

**Comment by fevac** [comment]
sorry copying the info here so @rannick can follow up (we chatted on slack):
The error is in the get_multiqc_search_patterns  you need to remove that definition from raredisease api to be able to use the general one.
That function returns a dictionary of search pattern in which the keys are the values that will be search in your multiqc file and the values are the ID that will be given in the metrics deliverable file (which is either case ID or sample ID)
So: {“my_search_patter_in_the_multiqc_file” : “my_id_in_the_metrics_deliverables_file”}
The general method in the nf_analysis api, uses sample IDs for both. But in your PR is overwritten to use case ID as a search pattern in multiqc, which is not found. I’ve checked your multiqc and the pattern seems to be sample IDs
I believe the docstring of that method might not be clear, so please feel free to rewrite to make it clearer.
Also let’s add and additional assert in test_metrics_deliver_case  to catch these errors. So make sure that metrics are actually written (and it’s not an empty list)

**Comment by peterpru** [comment]
Thanks @fevac and @rannick , I also see I need to fully complete the yaml file, which I will add next.

**Comment by fevac** [comment]
also, @peterpru don't forget to add raredisease to the `test_create_metrics_deliverables_content` test in cg/test/meta/workflow/test_nf_analysis.py

**Comment by peterpru** [comment]
It now appropriately populates the file.

image

**Comment by peterpru** [comment]
It looks like the command does what it should do when the sample relationships are correct. However, when the sample names are not correct in the multiqc-data.jason, the command does not fail, it just returns an empty list in the metrics_deliverables.yaml file. Example below. The output of the json fixture in the PR is also metrics: []

image

**Comment by peterpru** [comment]
When using the, `get_multiqc_search_patterns` from nf_analysis the yaml file gets created as expected:
image

However, the pytest `test_metrics_deliver_case` fails.

File: cg/meta/workflow/raredisease.py [hunk 10]
```diff
@@ -1,25 +1,27 @@
...
 )
-from cg.store.models import CaseSample
+from cg.models.nf_analysis import WorkflowParameters
+from cg.store.models import CaseSample, Sample
 
 LOG = logging.getLogger(__name__)
 
 
-class RarediseaseAnalysisAPI(NfAnalysisAPI):
+class RarediseaseAnalysisAPI(NfAnalysisAPI, ABC):
```
Comment (ivadym) [inline cg/meta/workflow/raredisease.py:24]: I don't think we need an abstract class here 🤔 

```suggestion
class RarediseaseAnalysisAPI(NfAnalysisAPI):
```

Comment (ChrOertlin) [inline cg/meta/workflow/raredisease.py:24]:
```suggestion
class RarediseaseAnalysisAPI(NfAnalysisAPI):
```
Why is this needed here?

File: tests/conftest.py [hunk 11]
```diff
@@ -2353,6 +2353,25 @@ def raredisease_nexflow_config_file_path(raredisease_dir, raredisease_case_id) -
...
+def raredisease_deliverable_data(
+    raredisease_dir: Path, raredisease_case_id: str, sample_id: str
+) -> dict:
+    return {
+        "files": [
+            {
+                "path": f"{raredisease_dir}/{raredisease_case_id}/multiqc/multiqc_report.html",
+                "path_index": "",
+                "step": "report",
+                "tag": ["multiqc-html", "rna"],
```
Comment (ChrOertlin) [inline tests/conftest.py:2366]:
```suggestion
                "tag": ["multiqc-html", "dna"],
```

**Review by ChrOertlin — APPROVED** [review]

File: tests/conftest.py (hunk 11 above)
Comment (fevac) [inline tests/conftest.py:2366]:
```suggestion
                "tag": ["multiqc-html"],
```

I think it's only "multiqc-html" for DNA cases

File: tests/meta/workflow/test_nf_analysis.py [hunk 12]
```diff
@@ -15,14 +15,14 @@
...
-    [Workflow.RNAFUSION, Workflow.TAXPROFILER, Workflow.TOMTE],
+    [Workflow.RAREDISEASE, Workflow.RNAFUSION, Workflow.TAXPROFILER, Workflow.TOMTE],
 )
 def test_create_metrics_deliverables_content(
     workflow: Workflow,
     caplog: LogCaptureFixture,
     request: FixtureRequest,
 ):
-    """Test metrics deliverables file content function for Taxprofiler and Rnafusion."""
+    """Test metrics deliverables file content function for nf-core pipelines."""
```
[the remark this answers is not in the document]
Comment (fevac) [reply tests/meta/workflow/test_nf_analysis.py:25]: True! Yes, should be nextflow

File: cg/models/raredisease/raredisease.py [hunk 13]
```diff
@@ -1,9 +1,18 @@
...
+class RarediseaseQCMetrics(BaseModel):
+    """Raredisease QC metrics"""
+
+    percentage_mapped_reads: float | None
+    PCT_TARGET_BASES_10X: float | None
+    MEDIAN_TARGET_COVERAGE: float | None
+    PCT_PF_READS_ALIGNED: float | None
+    PCT_PF_READS_IMPROPER_PAIRS: float | None
+    PCT_ADAPTER: float | None
+    fraction_duplicates: float | None
```
[the remark this answers is not in the document]
Comment (peterpru) [reply cg/models/raredisease/raredisease.py:15]: Ok, will add to next commit 👍

File: cg/meta/workflow/raredisease.py (hunk 10 above)
Comment (peterpru) [reply cg/meta/workflow/raredisease.py:24]: Ok I removed it in the next commit.

File: cg/models/raredisease/raredisease.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (peterpru) [reply cg/models/raredisease/raredisease.py:15]: ok thanks, adding to the next commit 👍

File: cg/meta/workflow/raredisease.py (hunk 10 above)
Comment (peterpru) [reply cg/meta/workflow/raredisease.py:24]: Removed 👍

File: tests/conftest.py (hunk 11 above)
Comment (peterpru) [reply tests/conftest.py:2366]: I'll go with Eva's suggestion below. 👍

Comment (peterpru) [reply tests/conftest.py:2366]: Added in next commit 👍

File: tests/conftest.py [hunk 14]
```diff
@@ -2272,7 +2272,7 @@ def sequencing_platform() -> str:
 
 # Raredisease fixtures
 @pytest.fixture(scope="function")
-def raredisease_dir(tmpdir_factory: Path) -> str:
+def raredisease_dir(tmpdir_factory, apps_dir: Path) -> str:
```
[the remark this answers is not in the document]
Comment (peterpru) [reply tests/conftest.py:2275]: removed 👍

**Comment by peterpru** [comment]
Deployed on stage:
image

**Comment by peterpru** [comment]
Deployed in production:
image
image
